proxy/src/control/discovery.rs

```diff
@@ -499,7 +526,121 @@ where T: HttpService<RequestBody = BoxBody, ResponseBody = RecvBody>,
     }
 }
 
-// ===== impl RxError =====
+// ===== impl DstLabels ====
+
+impl DstLabels {
+
+    #[inline]
+    fn new<I, S>(labels: I) -> Option<Self>
+    where
+        I: IntoIterator<Item=(S, S)>,
+        S: fmt::Display,
+    {
+        let mut labels = labels.into_iter();
+
+        if let Some((k, v)) = labels.next() {
+            // format the first label pair without a leading comma, since we
+            // don't know where it is in the output labels at this point.
+            let mut s = format!("dst_{}=\"{}\"", k, v);
+
+            // format subsequent label pairs with leading commas, since
+            // we know that we already formatted the first label pair.
+            for (k, v) in labels {
+                write!(s, ",dst_{}=\"{}\"", k, v)
+                    .expect("writing to string should not fail");
+            }
+
+            Some(DstLabels(Arc::from(s)))
+        } else {
+            // the iterator is empty; return None.
+            None
+        }
+
+    }
+
+}
+
+impl fmt::Display for DstLabels {
+
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+
+}
+
+// ===== impl Labeled =====
+
+impl Labeled<SocketAddr> {
+
+    /// Construct a new labeled `SocketAddr `from a protobuf `WeightedAddr`.
+    fn from_pb(pb: WeightedAddr, set_labels: &Arc<HashMap<String, String>>)
+               -> Option<Self> {
+        let inner = pb.addr.and_then(pb_to_sock_addr)?;
+        let label_iter =
+            set_labels.as_ref()
+                .iter()
+                .chain(pb.metric_labels.iter());
+        let metric_labels = DstLabels::new(label_iter);
+        Some(Labeled { metric_labels, inner, })
+    }
+}
+
+
+impl<T> Labeled<T> {
+    /// Wrap `inner` with no `metric_labels`.
+    pub fn none(inner: T) -> Self {
+        Self { metric_labels: None, inner }
+    }
+
+    /// Construct a new `Labeled<U>` for `inner` with the same labels as `self`.
+    fn label<U>(&self, inner: U) -> Labeled<U> {
+        Labeled {
+            metric_labels: self.metric_labels.as_ref().cloned(),
+            inner,
+        }
+    }
+}
+
+impl<T> ops::Deref for Labeled<T> {
```

Comment:
It is generally bad to implement `Deref` for any type that isn't a smart pointer. IIUC, the purpose of this implementation of `Deref` is to enable automatic conversions from `&Labeled<T>` to `&T`. However, as we discussed on IRC, such automatic conversions are often problematic, as shown in the `initial.retain(|x| other.contains(x)` example. The dangers of implementing `Deref` are well-documented in the Rust language documentation itself, which warns against doing it.

I understand that we have wanted such an automatic conversion because most things don't care about the labels, and so we don't want to change every type signature from `&T` to `Labeled<T>` or equivalent. With that in mind, it seems better to me to have `Labeled` trait that any labeled type of object can implement, that has a method that returns the labels (in some way). That way, we wouldn't feel any temptation to need to implement `Deref` for Labeled.

Reply:
I think that once we  modify the cache so that labels live as metadata in a map, this will no longer be an issue.

Reply:
I agree w/ @briansmith in this case with implementing `Deref` (though I am less strict about it than some).

Given the context that I have, I would lean against implementing a trait and instead just calling `get_ref()` on the labeled type.

Reply:
@carllerche I'll remove the `Deref` impl.